webrecorder/browsertrix
# Refactor Invites and Registration, Flatten Per-User Invites

Fixes #1432

This PR refactors the invite system to be simpler and more consistent with regards to existing user invites. Previously, per-user invites are stored in the user.invites dict instead of in the invites collection, which creates a few issues:
- Existing user do not show up in Org Invites list: #1432 
- Existing user invites also do not expire, unlike new user invites, creating potential security issue.

Instead, existing user invites should be treated like new user invites. This PR moves them into the same collection,
adding a `userid` field to InvitePending to match with an existing user.

If a user already exists, it will be matched by userid, instead of by email. This allows for user to update their email while still being invited. Note that the email of the invited existing user will not change in the invite email. This is also by design: an admin of one org should not be given any hint that an invited user already has an account, such as by having their email automatically update. For an org admin, the invite to a new or existing user should be indistinguishable.

The sha256 of invite token is stored instead of actual token for better security.

The registration system has also been refactored with the following changes:
- Auto-creation of new orgs for new users has been removed
- User.create_user() replaces the old User._create() and is much simpler
- Users are added to org in org add_user_to_org()
- Users are added to org through invites with add_user_with_invite()

Tests:
- Additional tests include verifying that existing and new pending invites appear in the pending invites list
- Tests for `/users/invite/<token>?email=` and `/users/me/invite/<token>` endpoints
- Deleting pending invites
- Additional tests added for user self-registration, including existing user self-registration to default org of existing user (in nightly tests)

Manual Testing:
- To test manually, create invites for new and existing users
- After running migration, invites for existing users should appear in the org list
- Can also try with self-registration enabled, registering as existing user to default org (lower priority)
